client/src/controllers/RulesController.test.js

```diff
@@ -0,0 +1,440 @@
+
+    it('should support an entries style array of key/value pairs to be used as an object', async () => {
+      await setup(`
+    <form data-controller="w-rules" data-action="change->w-rules#resolve">
+      <input type="text" name="title" value="bad" />
+      <input type="text" name="subtitle" value="bad" />
+      <input type="checkbox" name="agreement" id="agreement" checked />
+      <textarea
+        id="signature"
+        data-w-rules-target="enable"
+        data-w-rules="${_([
+          ['title', 'bad'],
+          ['subtitle', ['bad']],
+          ['agreement', ''],
+        ])}"
+        >
+      </textarea>
+    </form>
+      `);
+
+      expect(document.getElementById('signature').disabled).toBe(true);
+
+      document.getElementById('agreement').checked = false;
+      document
+        .getElementById('agreement')
+        .dispatchEvent(new CustomEvent('change', { bubbles: true }));
+
+      await jest.runAllTimersAsync();
+
+      expect(document.getElementById('signature').disabled).toBe(false);
+    });
+
+    it('should treat false/null as their string equivalents', async () => {
+      // Note: We may revisit this in the future, for now we want to confirm all values are resolved to strings for consistency.
+
+      await setup(`
+    <form data-controller="w-rules" data-action="change->w-rules#resolve">
+      <input id="confirm" type="checkbox" name="confirm" value="false" checked />
+      <input id="signUp" type="text" name="signUp" value="null" />
+      <input
+        class="test"
+        type="text"
+        name="a"
+        data-w-rules-target="enable"
+        data-w-rules='${_({ confirm: false })}' />
+      <input
+        class="test"
+        type="text"
+        name="b"
+        data-w-rules-target="enable"
+        data-w-rules="${_({ signUp: null })}" />
+    </form>
+      `);
+
+      // set to disabled when initialising based on rules
+      expect(
+        Array.from(document.querySelectorAll('.test')).map(
+          (target) => target.disabled,
+        ),
+      ).toEqual([false, false]);
+
+      document.getElementById('confirm').checked = false;
+      document
+        .getElementById('confirm')
+        .dispatchEvent(new CustomEvent('change', { bubbles: true }));
+
+      document.getElementById('signUp').value = '';
+      document
+        .getElementById('confirm')
+        .dispatchEvent(new CustomEvent('change', { bubbles: true }));
+
+      await jest.runAllTimersAsync();
+
+      expect(
+        Array.from(document.querySelectorAll('.test')).map(
+          (target) => target.disabled,
+        ),
+      ).toEqual([true, true]);
+    });
+
+    it('should support multiple values for the same field name correctly', async () => {
+      await setup(`
+    <form data-controller="w-rules" data-action="change->w-rules#resolve">
+      <fieldset>
+        <legend>Choose your favorite starship</legend>
+
+        <input type="radio" id="NCC-1701" name="starship" value="NCC-1701" />
```

Comment:
Would `type="checkbox"` be more appropriate for this test? You can't multi-select radio buttons, can you?

Reply:
Good call. I'll fix this up.